pkg/rosa/runtime.go

```diff
@@ -0,0 +1,71 @@
+package rosa
+
+import (
+	"os"
+
+	"github.com/openshift/rosa/pkg/aws"
+	"github.com/openshift/rosa/pkg/logging"
+	"github.com/openshift/rosa/pkg/ocm"
+	"github.com/openshift/rosa/pkg/reporter"
+	"github.com/sirupsen/logrus"
+)
+
+type Runtime struct {
+	Reporter  *reporter.Object
+	Logger    *logrus.Logger
+	OCMClient *ocm.Client
+	AWSClient aws.Client
+	Creator   *aws.Creator
+}
+
+func NewRuntime() *Runtime {
+	reporter := reporter.CreateReporterOrExit()
+	logger := logging.CreateLoggerOrExit(reporter)
+	return &Runtime{Reporter: reporter, Logger: logger}
+}
+
+// Adds an OCM client to the runtime. Requires a deferred call to `.Cleanup()` to close connections.
+func (r *Runtime) WithOCM() *Runtime {
+	if r.OCMClient == nil {
+		r.OCMClient = ocm.CreateNewClientOrExit(r.Logger, r.Reporter)
+	}
+	return r
+}
+
+// Adds an AWS client to the runtime
+func (r *Runtime) WithAWSClient() *Runtime {
+	if r.AWSClient == nil {
+		r.AWSClient = aws.CreateNewClientOrExit(r.Logger, r.Reporter)
+	}
+	return r
+}
+
+// Adds an AWS creator to the runtime. If the AWS client isn't initialized yet it will be.
+func (r *Runtime) WithAWSCreator() *Runtime {
+	if r.Creator != nil {
+		return r
+	}
+	if r.AWSClient == nil {
+		r = r.WithAWSClient()
+	}
+	var err error
+	r.Creator, err = r.AWSClient.GetCreator()
+	if err != nil {
+		r.Reporter.Errorf("Failed to get AWS creator: %v", err)
+		os.Exit(1)
+	}
+	return r
+}
+
+// Adds all clients to the runtime. Requires a deferred call to `.Cleanup()` to close connections.
+func (r *Runtime) WithEverything() *Runtime {
+	return r.WithOCM().WithAWSClient().WithAWSCreator()
+}
+
+func (r *Runtime) Cleanup() {
+	if r.OCMClient != nil {
+		if err := r.OCMClient.Close(); err != nil {
+			r.Reporter.Errorf("Failed to close OCM connection: %v", err)
+		}
```

Comment:
Can this be done inside `WithOCM`?
```

func (r *Runtime) WithOCM() *Runtime {
	if r.OCMClient == nil {
		r.OCMClient = ocm.CreateNewClientOrExit(r.Logger, r.Reporter)
	}
	defer r.OCMClient.Close()
	return r
}
```

Reply:
unfortunately not, as the defered method call will run when this function returns, not the caller